### ISSUE TITLE
DtypeWarning: Columns (0) have mixed types. 

### ISSUE THREAD

**Issue by ktroule** [griffithlab/pVACtools#149]
Sometimes I get this error message, I think it's related to pandas. 
Not sure if you've seen it before around here.
```
/home/.local/lib/python3.5/site-packages/lib/pipeline.py:347: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  self.binding_filter()
```
Thanks

**Comment by susannasiebert** [comment]
Does this make the run fail or is it just a warning?

Also, what version of pVACtools are you running?

**Comment by ktroule** [comment]
It seems that it works, one of the vcf files with this message outputs the final file with all peptides having at least one interaction below the IC50 threshold. 
I've seen this message previously when using pandas.

Just in case you didn't notice it, might be worth looking, not sure.
pvactools 1.0.5

**Comment by ktroule** [comment]
Yet three new issues. 
I'm working with two variant VCFs, in a server using queue (qsub). For each of the queue I reserve 4GB of memory.  Sometimes I get this message, somehow related to memory usage.
`Traceback (most recent call last):
  File "/home/ktroule/Programs/pVACtools-master/tools/pvacseq/run.py", line 159, in <module>
    main()
  File "/home/ktroule/Programs/pVACtools-master/tools/pvacseq/run.py", line 126, in main
    pipeline.execute()
  File "/home/ktroule/.local/lib/python3.5/site-packages/lib/pipeline.py", line 346, in execute
    self.combined_parsed_outputs(split_parsed_output_files)
  File "/home/ktroule/.local/lib/python3.5/site-packages/lib/pipeline.py", line 252, in combined_parsed_outputs
    '--top-score-metric', self.top_score_metric,
  File "/home/ktroule/.local/lib/python3.5/site-packages/lib/combine_parsed_outputs.py", line 42, in main
    for row in reader:
  File "/usr/lib/python3.5/csv.py", line 118, in __next__
    d = dict(zip(self.fieldnames, row))
MemoryError`

Other error, which seems related to making lots of calls to a database (www.cbs.dtu.dk), at first I thought it was related to the option  `--net-chop-threshold`, but after removing it it keeps failing, as the link expires I've not been able to see what is this calling to.
`Traceback (most recent call last):
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/connection.py", line 171, in _new_conn
    (self._dns_host, self.port), self.timeout, **extra_kw)
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/util/connection.py", line 79, in create_connection
    raise err
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/util/connection.py", line 69, in create_connection
    sock.connect(sa)
TimeoutError: [Errno 110] Connection timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/connectionpool.py", line 354, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/lib/python3.5/http/client.py", line 1107, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python3.5/http/client.py", line 1152, in _send_request
    self.endheaders(body)
  File "/usr/lib/python3.5/http/client.py", line 1103, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python3.5/http/client.py", line 934, in _send_output
    self.send(msg)
  File "/usr/lib/python3.5/http/client.py", line 877, in send
    self.connect()
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/connection.py", line 196, in connect
    conn = self._new_conn()
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/connection.py", line 180, in _new_conn
    self, "Failed to establish a new connection: %s" % e)
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x2b82cb215c50>: Failed to establish a new connection: [Errno 110] Connection timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ktroule/.local/lib/python3.5/site-packages/requests/adapters.py", line 445, in send
    timeout=timeout
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/connectionpool.py", line 638, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/util/retry.py", line 398, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='www.cbs.dtu.dk', port=80): Max retries exceeded with url: /cgi-bin/webface2.fcgi?jobid=5B48AE2300002B6484037726&wait=20 (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x2b82cb215c50>: Failed to establish a new connection: [Errno 110] Connection timed out',))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ktroule/Programs/pVACtools-master/tools/pvacseq/run.py", line 159, in <module>
    main()
  File "/home/ktroule/Programs/pVACtools-master/tools/pvacseq/run.py", line 126, in main
    pipeline.execute()
  File "/home/ktroule/.local/lib/python3.5/site-packages/lib/pipeline.py", line 376, in execute
    self.call_netmhc_stab()
  File "/home/ktroule/.local/lib/python3.5/site-packages/lib/pipeline.py", line 320, in call_netmhc_stab
    self.netmhc_stab_out_path(),
  File "/home/ktroule/.local/lib/python3.5/site-packages/lib/netmhc_stab.py", line 92, in main
    response = requests.get(response.url)
  File "/home/ktroule/.local/lib/python3.5/site-packages/requests/api.py", line 72, in get
    return request('get', url, params=params, **kwargs)
  File "/home/ktroule/.local/lib/python3.5/site-packages/requests/api.py", line 58, in request
    return session.request(method=method, url=url, **kwargs)
  File "/home/ktroule/.local/lib/python3.5/site-packages/requests/sessions.py", line 512, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/ktroule/.local/lib/python3.5/site-packages/requests/sessions.py", line 622, in send
    r = adapter.send(request, **kwargs)
  File "/home/ktroule/.local/lib/python3.5/site-packages/requests/adapters.py", line 513, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='www.cbs.dtu.dk', port=80): Max retries exceeded with url: /cgi-bin/webface2.fcgi?jobid=5B48AE2300002B6484037726&wait=20 (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x2b82cb215c50>: Failed to establish a new connection: [Errno 110] Connection timed out',))`

And the last, I think is also related to making lots of calls to a database, but I've not being able to figure exactly whats going on.
`Traceback (most recent call last):
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/connectionpool.py", line 384, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/connectionpool.py", line 380, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.5/http/client.py", line 1198, in getresponse
    response.begin()
  File "/usr/lib/python3.5/http/client.py", line 297, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.5/http/client.py", line 258, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib/python3.5/socket.py", line 576, in readinto
    return self._sock.recv_into(b)
ConnectionResetError: [Errno 104] Connection reset by peer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ktroule/.local/lib/python3.5/site-packages/requests/adapters.py", line 445, in send
    timeout=timeout
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/connectionpool.py", line 638, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/util/retry.py", line 367, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/packages/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/connectionpool.py", line 384, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "/home/ktroule/.local/lib/python3.5/site-packages/urllib3/connectionpool.py", line 380, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.5/http/client.py", line 1198, in getresponse
    response.begin()
  File "/usr/lib/python3.5/http/client.py", line 297, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.5/http/client.py", line 258, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib/python3.5/socket.py", line 576, in readinto
    return self._sock.recv_into(b)
urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ktroule/Programs/pVACtools-master/tools/pvacseq/run.py", line 159, in <module>
    main()
  File "/home/ktroule/Programs/pVACtools-master/tools/pvacseq/run.py", line 126, in main
    pipeline.execute()
  File "/home/ktroule/.local/lib/python3.5/site-packages/lib/pipeline.py", line 376, in execute
    self.call_netmhc_stab()
  File "/home/ktroule/.local/lib/python3.5/site-packages/lib/pipeline.py", line 320, in call_netmhc_stab
    self.netmhc_stab_out_path(),
  File "/home/ktroule/.local/lib/python3.5/site-packages/lib/netmhc_stab.py", line 92, in main
    response = requests.get(response.url)
  File "/home/ktroule/.local/lib/python3.5/site-packages/requests/api.py", line 72, in get
    return request('get', url, params=params, **kwargs)
  File "/home/ktroule/.local/lib/python3.5/site-packages/requests/api.py", line 58, in request
    return session.request(method=method, url=url, **kwargs)
  File "/home/ktroule/.local/lib/python3.5/site-packages/requests/sessions.py", line 512, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/ktroule/.local/lib/python3.5/site-packages/requests/sessions.py", line 622, in send
    r = adapter.send(request, **kwargs)
  File "/home/ktroule/.local/lib/python3.5/site-packages/requests/adapters.py", line 495, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))` 

once more, thanks for your time.

**Comment by ktroule** [comment]
Regarding the second error,I knew `net-chop` was call to a DB, but I've also seen that `net-stab` is also calling to a DB. 

Not sure if it's possible, but I think it will be, very interesting, to call those DB from local.

I'll investigate whether if I can get the stability after running 

Thanks

**Comment by susannasiebert** [comment]
In regards to the memory error, this happens when the combined.parsed.tsv file gets written, which contains all epitopes for all alleles, so the number of entries is significantly larger than the original VCF. This issue has been documented in #60. We're currently loading all of the tsv entries into memory in order to sort them, which is not memory-intensive and not ideal.

**Comment by susannasiebert** [comment]
The connection error you are seeing is for the `--netmhc-stab`. I'm not sure if there website is having intermittent connectivity problems or if this is a larger problem with trying to process a larger number of entries.

**Comment by ktroule** [comment]
Regarding the memory error, I'm currently just working with a 2 variant VCF file, surprised it requieres more than 4GB.

Once more, thanks for your kind help.

**Comment by susannasiebert** [comment]
Yes, that is indeed strange. Are those variants both frameshifts? Because that's the only reason I can imagine that the resulting file size would be that big. If that is the case you can try reducing the length of the downstream tail by setting the `--downstream-sequence-length` to something smaller .

**Comment by ktroule** [comment]
I'll check  it. I'm using a  `--downstream-sequence-length 100`, well below default.

**Comment by susannasiebert** [comment]
Hm, that seems to be plenty small, if you are able to share your VCF and the pvacseq command you are running, I'd be happy to help debug this.

**Comment by ktroule** [comment]
Once I see this I'll share it with you (I've delete all outputs to run it again)

thanks

**Comment by ktroule** [comment]
For instance, one of the VCF files that gives me error.
Traceback (most recent call last):
```
  File "/home/User1/Programs/pVACtools-master/tools/pvacseq/run.py", line 159, in <module>
    main()
  File "/home/User1/Programs/pVACtools-master/tools/pvacseq/run.py", line 126, in main
    pipeline.execute()
  File "/home/User1/.local/lib/python3.5/site-packages/lib/pipeline.py", line 346, in execute
    self.combined_parsed_outputs(split_parsed_output_files)
  File "/home/User1/.local/lib/python3.5/site-packages/lib/pipeline.py", line 252, in combined_parsed_outputs
    '--top-score-metric', self.top_score_metric,
  File "/home/User1/.local/lib/python3.5/site-packages/lib/combine_parsed_outputs.py", line 42, in main
    for row in reader:
  File "/usr/lib/python3.5/csv.py", line 110, in __next__
    row = next(self.reader)
MemoryError
```
Command line I use (3GB of memory assigned)
`python3.5 /home/User1/Programs/pVACtools-master/tools/pvacseq/run.py "$vcf" "$name" "$hla" NetMHCcons "$out" -e 9 -m lowest --iedb-install-directory /home/User1/Programs/ -b 1000 -c 1 --fasta-size 500 --downstream-sequence-length 100`

Attached the VCF file
[CosmicCodingMuts_Samples_Test_628.vcf.zip](https://github.com/griffithlab/pVACtools/files/2208792/CosmicCodingMuts_Samples_Test_628.vcf.zip)

**Comment by susannasiebert** [comment]
Unfortunately, I'm not able to reproduce this problem locally. We plan on rewriting this module to be less memory-intensive. Please follow https://github.com/griffithlab/pVACtools/issues/60 to monitor the progress on this issue.

**Comment by susannasiebert** [comment]
The PR (#151) to remove the `DtypeWarning: Columns (0) have mixed types.` warning has been merged in and released as part of hotfix release 1.0.8. Please update to the latest version to see this change. I'm resolving this issue.

**Comment by ktroule** [comment]
Thanks.
